File: Python/trnc_holiday_validator.py
from datetime import datetime
from HolidayChecker.HolidayValidator import HolidayValidator

class TRNCHolidayValidator(HolidayValidator):
    """
    TRNCHolidayValidator

    This class is used to determine if a given date is a holiday in TRNC.
    """

    FIXED_HOLIDAYS = HolidayValidator.FIXED_HOLIDAYS + [
        '08-01',  # TMT Günü
        '11-15',  # Kuzey Kıbrıs Türk Cumhuriyeti'nin İlanı
    ]

    BASE_DATES = HolidayValidator.BASE_DATES + [
        {'name': 'mevlidKandili', 'date': '2024-09-15', 'days': 1}
    ]

    def __init__(self, include_saturday=False):
        """
        Initialize the TRNCHolidayValidator

        :param include_saturday: Boolean flag indicating whether to include Saturdays as holidays
        """
        super().__init__(include_saturday)

File: C++/src/TRNCHolidayValidator.cpp
#include "TRNCHolidayValidator.h"
#include <algorithm>

const std::vector<std::string> TRNCHolidayValidator::trncFixedHolidays = {
    "08-01", // TMT Günü
    "11-15"  // Kuzey Kıbrıs Türk Cumhuriyeti'nin İlanı
};

const std::vector<std::tuple<std::string, std::tm, int>> TRNCHolidayValidator::trncBaseDates = {
    {"mevlidKandili", {0, 0, 0, 15, 8, 124}, 1} // 2024-09-15
};

TRNCHolidayValidator::TRNCHolidayValidator(bool includeSaturday) : HolidayValidator(includeSaturday) {}

bool TRNCHolidayValidator::isFixedHoliday(const std::tm &date) {
    char buffer[6];
    std::strftime(buffer, sizeof(buffer), "%m-%d", &date);
    return std::find(trncFixedHolidays.begin(), trncFixedHolidays.end(), buffer) != trncFixedHolidays.end() ||
           HolidayValidator::isFixedHoliday(date);
}

std::vector<std::tuple<std::string, std::tm, int>> TRNCHolidayValidator::getBaseDates() {
    auto baseDates = HolidayValidator::baseDates;
    baseDates.insert(baseDates.end(), trncBaseDates.begin(), trncBaseDates.end());
    return baseDates;
}
